sources/Messages/Message.php

```diff
@@ -29,6 +29,11 @@ public function __construct
         $this->body = $body;
         $this->email = $email;
         $this->status = $status;
+        $this->id = 0;
+        $this->subject = '';
+        $this->phone = '';
+        $this->address = new Address('', 0, '', '', '');
```

Comment:
Acredito que será melhor definir `null`, visto que fica difícil diferenciar quando realmente existe ou não um endereço.

Reply:
O problema de definir null seria na hora do bind realizado lá no insert. Teria que ser feita uma condição para ver se Message::getAddress retornaria algo, se não retornasse, não seria feitos os binds com os campos de endereço, daí teria que ser alterado também os campos retornados no método Sql::getFields() @johnnysaymon

Reply:
Concorda comigo que esses motivos não favorecem o uso da aplicação, apenas facilita em alguns pontos no desenvolvimento dela mesma? 
Os locais onde ele irá atrapalhar é justamente em pontos axilares que foram produzidos para facilitar, sendo assim é bom revermos se optamos por um bom caminho.

Reply:
Confesso que não consigo ver pela sua ótica. Mas posso alterar o retorno para null e procurar outra saída nos binds @johnnysaymon

Reply:
Acredito que ficará mais claro se você tentar enxergar a aplicação de fora, como quem irá consumir, que não precisa e nem deve saber como as coisas ocorrem internamente. Para quem consome basta saber quais os métodos que salva no *Storage* e conhecer as possibilidades com os objetos *Message* e *Address*, desse ponto de vista não existe nenhum motivo para *Message* retornar um objeto *Address* quando ele não existe.

Porém, devo confessar que gostaria de retornar sempre um objeto e nunca *null*, porém se fosse evitar isso iria criar uma *inferface* Address e duas classes, uma *AddressConcrect* que é a que você já tem e *AddressUndefined* que representaria a falta de um endereço, também iria _tipar_ o retorno de *Message* com a interface e não com a classe concreta.

Reply:
@johnnysaymon , essa classe AddressUndefined iria ter as mesmas propriedades que AddressConcret ou não existiriam propriedades?

Reply:
Não, será uma classe vazia, sem nenhuma propriedade ou método.

Reply:
@johnnysaymon , o que você acha de ter um método Sql::getFields e outro Sql::getFieldsWithoutAddress. A mesma coisa para o método Sql::getBinds

Reply:
Considero a criação do método `getFieldsWithoutAddress` a abertura para uma criação de outras infinidades de combinações possíveis, o que fará a classe se encher de métodos que fazem quase a mesma coisa, sei que nesse caso não se haverá muitas outras possibilidades de métodos, mas ainda assim estamos criando um novo método para retornar quase as mesmas informações de `getFields`.

Sobre a criação de métodos com parâmetros booleanos, acredito que esse caso é uma exceção, visto que o método ainda cumpre a mesma função, ele continua retornando uma string com os campos, trata-se apenas de um filtro. Acredito que na palestra que viu ele deve estar se referindo a parâmetros booleanos que alteram o comportamento do método e que outros tipos de parâmetros poderiam cometer o mesmo erro, mas provavelmente ele deve ter se deparado mais com métodos que recebem booleanos.

Mas o método pode sim ser modificado, pois está surgindo a necessidade de aperfeiçoar o filtro, o que você poderá fazer é algo assim:

```php
private function getFields(string $aliasTable = null, bool $modeInclude, string ...$filter): string
```

Caso o `$modeInclude` seja `true` você só irá retornar o que estiver definido em `$filter`, caso seja `false` só não retorna o que estiver definido nos filtros.

Reply:
Aqui eu vejo um pequeno problema, caso o o `$modeInclude`, irá retornar o que estiver definido em `$filter`. Assim a fonte dos dados deixará de ser o retornado em getColumns. Os dados podem acabar não sendo confiáveis @johnnysaymon

Reply:
Não compreendi @maikenegreiros . Você está se referindo ao retorno padrão, quando não existir filtros?

Se for isso, realmente será preciso definir o valor padrão do parâmetro `$modeInclude` como `false`.

Reply:
Não @johnnysaymon , estou falando das colunas retornadas em getColumns. As colunas não mais virão desse método e sim do parâmetro ...$filter.

Reply:
Não, sua fonte continuará sendo `getColumns`, o que `getFields`irá fazer é filtrar os dados antes de retornar a string conforme configuração dos parâmetros @maikenegreiros

Reply:
Mas você disse que se `$modeInclude` for `true`, as colunas seriam as passadas no parâmetro `...$filters` @johnnysaymon

Reply:
Acredito que com exemplo ficará mais fácil compreender:

```php

// Obtendo todas as colunas sem alias
$this->getFields();

// Obtendo apena as colunas id e name com alias
$this->getFields('message', true, 'id', 'name');

// Obtendo todas menos o ID e com alias
$this->getFields('message', false, 'id');

// Obtendo todas menos o as relacionadas a endereço
$this->getFields('message', false, 'address'); # Essa vai ter que existir uma condição

// Obtendo todas menos o nome da cidade
$this->getFields('message', false, 'address_city_name')
```

Reply:
No segundo exemplo que você deu, as colunas que são desejadas estão sendo passadas por parâmetro. Concorda que nesse caso é desnecessário chamar getColumns @johnnysaymon ?

Reply:
Concordo. Sendo assim, podemos esquecer o parâmetro booleano e assumir que todos os campos passados devem ser os que precisam ser ignorados. Ou, ficarmos só com o quarto exemplo e não especificar os campos.